[#193] Expanded concept mappings capabilities.

mseaton opened the pull request:
Expand concepts domain to support concept mappings other than same-as, and to enable mappings to be specified in multiple columns if desired.

See #193 for details.

mks-d commented on api/src/main/java/org/openmrs/module/initializer/api/c/MappingsConceptLineProcessor.java:
```diff
@@ -4,36 +4,75 @@
+	protected final Logger log = LoggerFactory.getLogger(getClass());
 	
-	private ConceptMapListParser listParser;
+	public static final String MAPPING_HEADER_PREFIX = "mapping";
```
I would favour here to go for plural since those headers accept arrays of values:
```java
public static final String MAPPINGS_HEADER_PREFIX = "mappings";
```
and so one for associated headers.

mseaton replied:
@mks-d I have pushed a new commit that makes this change to "mappings" plural

ibacher commented on api/src/main/java/org/openmrs/module/initializer/api/c/MappingsConceptLineProcessor.java:
```diff
@@ -4,36 +4,75 @@
+					header = MAPPING_HEADER_PREFIX + MAPPING_HEADER_SEPARATOR + ConceptMapType.SAME_AS_MAP_TYPE_UUID;
+				}
+				
+				String[] headerComponents = header.split(MAPPING_HEADER_SEPARATOR_REGEX);
```
I think this is slightly preferable...

```suggestion
				String[] headerComponents = header.split(MAPPING_HEADER_SEPARATOR_REGEX, 3);
```

mseaton replied:
I had this originally, but changed it to be less rigid.  There are actually up to 4 components (mappings|type|source|unique-key), though only 2 components are required.

mseaton replied:
@ibacher I have pushed a new commit that limits the split to 4 components.

ibacher commented on api/src/main/java/org/openmrs/module/initializer/api/c/MappingsConceptLineProcessor.java:
```diff
@@ -4,36 +4,75 @@
+				if (headerComponents[0].equalsIgnoreCase(MAPPING_HEADER_PREFIX)) {
+					
+					// First determine the map type (required)
+					String headerMapTypeStr = headerComponents[1].trim().toLowerCase();
```
So, OMRS can be configured to run in either case-sensitive or case-insensitive mode for database searches; this seems to require OMRS be running in case-insensitive mode. Do we need this normalisation to lower case?

mseaton replied:
That's a good point.  I tend to take case-insensitive db searching as a given in openmrs.  We need the header to be case-insensitive I think to be consistent with every other header in iniz.  But defer to @mks-d .  I think this is likely an issue throughout the entire module, not just with this particular component.

ibacher reviewed (approved):
Looks great @mseaton! Two small things I might adjust, but I don't see any major issues with it.

brandones commented:
Updating the exporter: https://github.com/PIH/iniz-exporters/pull/4

mseaton commented:
I think I have addressed the PR comments.  Let me know if I'm good to merge this in or if this needs more review or one of you would prefer to merge it in when ready.  Thanks @mks-d and @ibacher
